setup.py

```diff
@@ -25,6 +25,7 @@
     python_requires=">=3.7",
     install_requires=[
         "python-dateutil>=2.7.0",
+        "ciso8601~=2.2.0",
```

Comment:
We've traditionally been *very* stingy about adding new dependencies to PySTAC. My feeling is that this one is worth it, but could we could drop `python-dateutil` if we are introducing `ciso8601`?

@gadomski @lossyrob @schwehr Curious to get your thoughts as well.

Reply:
To give a bit of context to this, ciso8601 is the only library I've found in any language that correctly parses RFC 3339 dates. I'd though about replacing python-dateutil instead, but that would likely break existing behavior because the python-dateutil parsing methods are *so* permissive in what they accept -- it's very much an attempt to convert anything that looks like a string representing a datetime into a `datetime.datetime`.

Reply:
Also, ciso8601 introduces no transitive dependencies, as it has none.

Reply:
I am slightly against it as a requirement.  

- [It's C code](https://github.com/closeio/ciso8601/blob/master/module.c)... so that's a weightier dependency to have
- It's not that commonly used
- But at least [it is MIT licensed](https://github.com/closeio/ciso8601/blob/master/LICENSE), so no license concerns from my point-of-view

If there was a way to have it be an optional install that didn't make the testing situation a nightmare, that would be good.

Reply:
I removed `str_to_interval`, since I think that is very specific to api impls and should be in stac-fastapi instead.

Reply:
I've been thinking about this more, and my underlying concern here is that the use of dateutil.parser.parse in `str_to_datetime` (which is used everywhere) is too lenient, and means that many datetime strings that are not even close to being RFC3339 are parsed without error. For example, `dateutil.parser.parse("January 1, 2047 at 8:21:00AM")` returns a valid datetime. Postel's law is relevant here, but as it is, I think pystac is being much too liberal in what it accepts as a datetime string. 

Another option would be to use `dateutil.parser.isoparse` that only parses ISO8601 datetimes. I haven't personally tested this against my edge case strings, but will do that.

Reply:
So there's two issues here:

1. should the existing str_to_datetime be changed to be more strict
2. should a precise RFC3339 method (useful for validation) be included? Should a correct datetime->str method be included for use by downstream users?

Reply:
> Curious to get your thoughts as well.

I'm with @schwehr, if we're going to incorporate a non-wheel-providing dependency that requires a compiler, I'd like to move slowly and ensure there aren't any Python-only alternative solutions.

To address your questions @philvarner, from my perspective:

> should the existing str_to_datetime be changed to be more strict

PySTAC is used a part of production systems (where we it to be conservative) and by data consumers in development/research environments (where we want it to be liberal). I'd be in favor of a opt-in strictness on the consumption side (i.e. `str -> datetime`) to ensure that our users in the science community (e.g.) don't suddenly have to wrangle w/ datetime parsing errors because the data provider they've been using has been Doing It Wrong™.

> should a precise RFC3339 method (useful for validation) be included?

Yes. Validation should be strict as hell right? :-)

> Should a correct datetime->str method be included for use by downstream users?

Yes. Production should be as strict as possible IMO.

> Another option would be to use dateutil.parser.isoparse that only parses ISO8601 datetimes. I haven't personally tested this against my edge case strings, but will do that.

IMO this would be a much more attractive alternative if we can make it work.

Reply:
IMO accepting datetimes liberally is a feature, not a bug. The purpose of PySTAC is to make working with STAC easier, not leave people scratching their heads on what the exact correct ISO8601 formatted string is to make PySTAC accept their datetime string (which I know I would be). Adding a more strict version of str_to_datetime does seem like a great addition; however like Pete mentions we should for sure not break existing usage of PySTAC in the name of strictness, even if we wanted to (which as I said, I wouldn't be in favor of strictness).

For the write site (datetime_to_str), that should always produce valid datetimes to the standard. Any issues solved around that is a good change. "Lenient on input, strict on output" is a good mantra for this library IMO.

I think there's likely a way to solve this that doesn't include the addition of a C dependency. I think we'd have to weight whatever additional capability against the addition of a dependency, which as others have said we've been pretty careful not to do. From what I can tell the work to fix issues with the datetime_to_str method in a Python-only way would be worth excluding this additional dependency and saving the potential work by the user base for any future issues that dependency may cause.

Reply:
1. rewrote to remove the pytest dependency. I should have just done it that way from the beginning. 
2. changed the new method to use dateutil's ISO8601 parsing method. Another option would be to add a parameter "strict" to `str_to_datetime` that defaults to False for now, and maybe change it to True in a future version with appropriate notification of the the change.

The strictness should probably be somewhere between accepting "January 1, 2047 at 8:21:00AM" and requiring a perfect RFC3339 string.

Reply:
The scope of this PR has been pared down dramatically since this conversation, to the point where it might end up just adding two free helper functions to `utils`. I'm going to consider this conversation useful but resolved.